### PR TITLE
 Enable users to use their own loss functions + deal with prefetching for grad accum

### LINKED ISSUE #2175: Gradient accumulation yields worse results than the equivalent batch size
I expected a training configuration with per_device_train_batch_size=1 and gradient_accumulation_steps=32 to yield the same (or similar) result to per_device_train_batch_size=32 and gradient_accumulation_steps=1 but that's not the case, the former is much worse.
I ran several experiments with SmolLM-135M and Llama 3.2 1B, using always the same seed, and the results are consistent with this observation.

image
image

Maybe I misunderstand something here?

My training code is in [this Colab notebook](https://colab.research.google.com/drive/17g7zVSvGragLiB6DlHGW-D3X0TyfxKHg?usp=sharing). I ran this notebook to draw the learning curves above, restarting the notebook between each training to avoid OOM.
Note that I have the same observations with Qwen2.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

In conjunction with https://github.com/huggingface/transformers/pull/34191, this PR solves the other half of what's needed:

1. Letting users pass in their own loss functions directly to the Trainer via `compute_loss`
2. Prefetching the first `gradient_accumulation_steps` worth of data each complete step and marking how many samples were seen (`num_items_in_batch`), which can be passed to a loss function if it takes in `num_items_seen` (name TBD)

A bit of feedback needed we need to coordinate:

* Should it be called `num_items_in_batch` and then passed through to the loss functions as such? Or is there a better name we can think of

Fixes https://github.com/huggingface/trl/issues/2175

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik @ArthurZucker

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2406,142 +2408,178 @@ def _inner_training_loop(
...
+                            self.state.num_input_tokens_seen += (
+                                torch.sum(
+                                    self.accelerator.gather(
+                                        torch.tensor(
+                                            inputs[main_input_name].numel(), device=self.args.device, dtype=torch.int64
+                                        )
                                     )
                                 )
+                                .cpu()
+                                .item()
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:2472]: let's make this more readable!

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3595,13 +3635,13 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
-    def compute_loss(self, model, inputs, return_outputs=False):
+    def _compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
         """
         How the loss is computed by Trainer. By default, all models return the loss in the first element.
 
         Subclass and override for custom behavior.
         """
-        if self.label_smoother is not None and "labels" in inputs:
+        if (self.label_smoother is not None or self.compute_loss is not None) and "labels" in inputs:
             labels = inputs.pop("labels")
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:3645]: mmmm if people don't pass a loss, we won't use the model's default?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -427,6 +427,7 @@ def __init__(
             Union[PreTrainedTokenizerBase, BaseImageProcessor, FeatureExtractionMixin, ProcessorMixin]
         ] = None,
         model_init: Optional[Callable[[], PreTrainedModel]] = None,
+        compute_loss: Optional[Callable] = None,
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:430]: needs to be documented somewhere!

**Review by ArthurZucker — COMMENTED** [review]
LGTM, IMO a regression test on the grad norms could be fairly nice!

**Comment by muellerzr** [comment]
A bit more context, full fine-tuning does **NOT SEEM TO BE IMPACTED BY THIS**. I am looking into how this directly affects TRL, however *things are not as bad as they may seem*. 

(Below is an example CausalLM result comparing grad accum 4, bs 8 vs bs 32 both before and after this fix)

image

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -2406,142 +2408,178 @@ def _inner_training_loop(
...
+                for _ in range(num_batches):
+                    try:
+                        batch_samples += [next(epoch_iterator)]
+                    except StopIteration:
+                        break
+                if len(batch_samples) > 0 and "labels" in batch_samples[0]:
+                    # For now we don't support object detection
+                    try:
+                        num_items_in_batch = sum(
+                            [data_batch["labels"][..., 1:].ne(-100).sum().item() for data_batch in batch_samples]
```
Comment (BenjaminBossan) [inline src/transformers/trainer.py:2441]: I already quickly discussed this with Zach, so this is a more general questions to other reviewers:

Would this line be work for all the different task types we support? Specifically, can we always skip the first item in the sequence, i.e. is the `[..., 1:]` part valid?

File: src/transformers/trainer.py (hunk 2 above)
Comment (muellerzr) [reply src/transformers/trainer.py:3646]: We will, it stays in `inputs` and gets passed to the models `forward()`

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -3591,20 +3631,23 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                 scaled_loss.backward()
         else:
+            loss *= self.args.gradient_accumulation_steps
             self.accelerator.backward(loss, **kwargs)
 
         return loss.detach() / self.args.gradient_accumulation_steps
 
-    def compute_loss(self, model, inputs, return_outputs=False):
+    def _compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
```
Comment (winglian) [inline src/transformers/trainer.py:3639]: aren't there a lot of downstream trainers that extend this?

Comment (winglian) [reply src/transformers/trainer.py:3639]: for example https://github.com/huggingface/trl/blob/a67f2143c38d6520be8735463ce715ad5c281db8/trl/trainer/cpo_trainer.py#L826

Comment (muellerzr) [reply src/transformers/trainer.py:3652]: I'll reverse this then (store as `_compute_loss` and keep `compute_loss`)

Comment (winglian) [reply src/transformers/trainer.py]: thanks, btw, @lewtun we should open a PR to add `num_items_in_batch` or `**kwargs` to `compute_loss` otherwise this will break trl too

File: src/transformers/trainer.py (hunk 4 above)
Comment (danielhanchen) [reply src/transformers/trainer.py]: For casual auto regressive models it works but won't work in other ones

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -168,6 +183,25 @@ def on_step_end(self, args, state, control, **kwargs):
...
+
+    # Flatten the tokens
+    shift_logits = shift_logits.view(-1, vocab_size)
+    shift_labels = shift_labels.view(-1)
+    # Enable model parallelism
+    shift_labels = shift_labels.to(shift_logits.device)
+    if num_items_in_batch is None or disable_num_items_in_batch:
+        num_items_in_batch = shift_labels.numel()
+    loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
+    loss = loss / num_items_in_batch
```
Comment (danielhanchen) [inline tests/trainer/test_trainer.py:201]: Although this is a test, it's probably better to not divide by `.numel()` or the number of items, but do the pure mean, since there might be padding tokens.

```suggestion
    if num_items_in_batch is None or disable_num_items_in_batch:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="mean")
    else:
        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
        loss = loss / num_items_in_batch
```

**Review by danielhanchen — CHANGES_REQUESTED** [review]
Just a denominator change in the test case

**Review by ArthurZucker — APPROVED** [review]
Feel free to merge!

File: src/transformers/trainer.py (hunk 5 above)
Comment (lewtun) [reply src/transformers/trainer.py]: thanks @winglian !
